generator.c

```diff
@@ -0,0 +1,122 @@
+#include "common.h"
+#include "generator.h"
+#include "iterative.h"
+#include "recursive.h"
+#include "singlethreaded.h"
+
+#include <pthread.h>
+#include <stdbool.h>
+#include <string.h>
+#include <alloca.h>
+#include <unistd.h>
+
+struct gn_context_t
+{
+    pthread_mutex_t mutex;
+    password_t password;
+    char *hash;
+    volatile bool found;
+    volatile bool done;
+
+    struct config_t *config;
+
+    union {
+        struct iter_state_t iter_state[0];
+        struct rec_state_t rec_state[0];
+    };
+};
+
+static void *
+gn_worker(void *arg)
+{
+    struct gn_context_t *context = (struct gn_context_t *) arg;
+    struct config_t *config = context->config;
+
+    struct st_context_t st_context;
+    st_context.hash = context->hash;
+    st_context.cd.initialized = 0;
+
+    while (true)
+    {
+        struct task_t task;
+        pthread_mutex_lock(&context->mutex);
+        bool done = context->done;
+        if (!done)
+        {
+            switch (config->brute_mode)
+            {
+            case M_ITERATIVE:
+                task = *context->iter_state->task;
+                context->done = !iter_next(context->iter_state);
+                break;
+            case M_RECURSIVE:
+            case M_REC_ITERATOR:
+                task = *context->rec_state->task;
+                context->done = !rec_next(context->rec_state);
+                break;
+            }
```

Comment:
For invalid brute_mode value cycle will be infinite. Need to add default branch for that case.